src/python/pants/init/engine_initializer.py

```diff
@@ -371,6 +374,19 @@ def union_membership_singleton() -> UnionMembership:
     def build_root_singleton() -> BuildRoot:
       return BuildRoot.instance
 
+    @rule
+    def single_build_file_address(specs: Specs) -> BuildFileAddress:
+      build_file_addresses = yield Get(BuildFileAddresses, Specs, specs)
+      if len(build_file_addresses.dependencies) != 1:
+        provenanced_addresses = yield Get(ProvenancedBuildFileAddresses, Specs, specs)
+        matched_addrs = [addr.build_file_address.to_address() for addr in provenanced_addresses]
+        output = '\n '.join(f':{addr}' for addr in matched_addrs)
+        raise ResolveError(
+          "This goal only works with a single target, but has been given multiple targets:\n"
```

Comment:
I'm not sure this is important, but I think readability by end users at the CLI remains high if you cut the non-local reference to "goal" which this rule is not. Perhaps:
```
"Expected a single target, but has been given multiple targets:\n"
f"Did you mean one of:\n {output}"
```